runtime/sema/type.go

```diff
@@ -5519,6 +5620,19 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
 			return typedSubType.ExplicitInterfaceConformanceSet().
 				Contains(typedSuperType)
 
+		// An interface type is a supertype of a restricted type if the restricted set contains
+		// that explicit interface type. Once interfaces can conform to interfaces, this should instead
+		// check that at least one value in the restriction set is a subtype of the interface supertype
+
+		// This particular case comes up when checking attachment access; enabling the following expression to typechecking:
+		// resource interface I { /* ... */ }
+		// attachment A for I { /* ... */ }
+		//
+		// let i : &{I} = ... // some operation constructing `i`
+		// let a = i[A] // must here check that `i`'s type is a subtype of `A`'s base type, or that {I} <: I
+		case *RestrictedType:
+			return typedSubType.RestrictionSet().Contains(typedSuperType)
```

Comment:
Same as above: Maybe this can be a separate function that is only used for this purpose.

The `IsSubType` function is already very complex, and so far assumes / could rely on the fact that interfaces only appear in restricted types. I'm worried the rules / this function becomes incomplete.

Reply:
https://github.com/onflow/cadence/pull/2099#discussion_r1042362870